src/rgw/rgw_compression.cc

```diff
@@ -0,0 +1,166 @@
+                                                                first_block(0),
+                                                                last_block(0),
+                                                                q_ofs(0),
+                                                                q_len(0),
+                                                                first_data(true),
+                                                                cur_ofs(0)
+{
+  compressor = Compressor::create(cct, cs_info->compression_type);
+  if (!compressor.get())
+    lderr(cct) << "Cannot load compressor of type " << cs_info->compression_type 
+                     << " for rgw, check rgw_compression_type config option" << dendl;
+}
+
+int RGWGetObj_Decompress::handle_data(bufferlist& bl, off_t bl_ofs, off_t bl_len)
+{
+  ldout(cct, 10) << "Compression for rgw is enabled, decompress part " << bl_len << dendl;
+
+  if (!compressor.get()) {
+    // if compressor isn't available - error, because cannot return decompressed data?
+    lderr(cct) << "Cannot load compressor of type " << cs_info->compression_type 
+                     << " for rgw, check rgw_compression_type config option" << dendl;
+    return -EIO;
+  }
+  bufferlist out_bl, in_bl;
+  bl_ofs = 0;
+  if (waiting.length() != 0) {
+    in_bl.append(waiting);
+    in_bl.append(bl);        
+    waiting.clear();
+  } else {
+    in_bl.claim(bl);
+  }
+  bl_len = in_bl.length();
+  
+  while (first_block <= last_block) {
+    bufferlist tmp, tmp_out;
+    int ofs_in_bl = cs_info->blocks[first_block].new_ofs - cur_ofs;
+    if (ofs_in_bl + cs_info->blocks[first_block].len > bl_len) {
+      // not complete block, put it to waiting
+      int tail = bl_len - ofs_in_bl;
+      in_bl.copy(ofs_in_bl, tail, waiting);
+      cur_ofs -= tail;
+      break;
+    }
+    in_bl.copy(ofs_in_bl, cs_info->blocks[first_block].len, tmp);
+    int cr = compressor->decompress(tmp, tmp_out);
+    if (cr < 0) {
+      lderr(cct) << "Compression failed with exit code " << cr << dendl;
+      return cr;
+    }
+    if (first_block == last_block && partial_content)
+      out_bl.append(tmp_out.c_str(), q_len);
+    else
+      out_bl.append(tmp_out);
+    first_block++;
+  }
+
+  if (first_data && partial_content && out_bl.length() != 0)
+    bl_ofs =  q_ofs;
+
+  if (first_data && out_bl.length() != 0)
+    first_data = false;
+
+  cur_ofs += bl_len;
+  return next->handle_data(out_bl, bl_ofs, out_bl.length() - bl_ofs);
+}
+
+void RGWGetObj_Decompress::fixup_range(off_t& bl_ofs, off_t& bl_len)
+{
+  if (partial_content) {
+    // if user set range, we need to calculate it in decompressed data
+    first_block = 0; last_block = 0;
+    if (cs_info->blocks.size() > 1) {
+      off_t i = 1;
+      while (i < cs_info->blocks.size() && cs_info->blocks[i].old_ofs <= bl_ofs) i++;
+      first_block = i - 1;
+      while (i < cs_info->blocks.size() && cs_info->blocks[i].old_ofs < bl_len) i++;
```

Comment:
we can use a binary search for these, right? `std::upper_bound()` for the first, and `std::lower_bound()` for the second?